### PR TITLE
uniq: avoid building list of duplicate lines

### PR DESCRIPTION
**Pull request by cbjadwani**

### CODE REVIEW

**Comment by sylvestre** [comment]
Could you please share more details on the why? thanks

**Comment by cbjadwani** [comment]
This reduces memory usage by only storing two lines of the input file at a time. The current implementation first builds a list of all duplicate lines ('group') and then decides which lines of the group should be printed.

Also added this in commit description.

File: src/uu/uniq/src/uniq.rs [hunk 1]
```diff
@@ -61,35 +61,42 @@ impl Uniq {
...
-                let print_delimiter = !no_delimiters && (prepend_delimiter || first_line_printed);
-                first_line_printed |= self.print_lines(writer, &lines, print_delimiter);
-                lines.truncate(0);
+        let mut lines = reader.split(line_terminator).map(get_line_string);
+        let mut line = match lines.next() {
+            Some(l) => l,
+            None => return,
+        };
+
+        for next_line in lines {
```
Comment (sylvestre) [inline src/uu/uniq/src/uniq.rs:73]: please a comment explaining what is loop is doing

**Review by sylvestre — COMMENTED** [review]
just a tiny nit

File: src/uu/uniq/src/uniq.rs [hunk 2]
```diff
@@ -61,35 +61,44 @@ impl Uniq {
...
-            // before or if we need to start output with delimiter
-            let print_delimiter = !no_delimiters && (prepend_delimiter || first_line_printed);
-            first_line_printed |= self.print_lines(writer, &lines, print_delimiter);
+        if (group_count == 1 && !self.repeats_only) || (group_count > 1 && !self.uniques_only) {
+            self.print_line(writer, &line, group_count, first_line_printed);
+            first_line_printed = true;
         }
-        if append_delimiter && first_line_printed {
-            crash_if_err!(1, writer.write_all(&[line_terminator]));
+        if self.delimiters == Delimiters::Append || self.delimiters == Delimiters::Both {
```
Comment (sylvestre) [inline src/uu/uniq/src/uniq.rs:98]: Sorry but could you please fix this warning?

**Pull request merged**